### PR TITLE
infa: add python fuzz-introspector support

### PR DESCRIPTION
**Pull request by DavidKorczynski**
Adds necessary logic for running fuzz-introspector's Python frontend.

Signed-off-by: David Korczynski <david@adalogics.com>

### CODE REVIEW

**Comment by DavidKorczynski** [comment]
Ref: https://github.com/ossf/fuzz-introspector/issues/280

**Review by Navidem — APPROVED** [review]
LGTM

**Comment by jonathanmetzman** [comment]
@Navidem OK I don't need trial builds now, please test this before landing.

File: infra/base-images/base-builder/compile [hunk 1]
```diff
@@ -42,7 +42,7 @@ if [ "$FUZZING_LANGUAGE" = "python" ]; then
     echo "ERROR: Python projects can be fuzzed with libFuzzer engine only."
     exit 1
   fi
-  if [ "$SANITIZER" != "address" ] && [ "$SANITIZER" != "undefined" ] && [ "$SANITIZER" != "coverage" ]; then
+  if [ "$SANITIZER" != "address" ] && [ "$SANITIZER" != "undefined" ] && [ "$SANITIZER" != "coverage" ] && [ "$SANITIZER" != "introspector" ]; then
     echo "ERROR: Python projects can be fuzzed with AddressSanitizer or UndefinedBehaviorSanitizer or Coverage only."
```
Comment (Navidem) [inline infra/base-images/base-builder/compile:46]: nit: Error message is missing Introspector as a valid sanitizer.

File: infra/base-images/base-builder/compile_python_fuzzer [hunk 2]
```diff
@@ -24,6 +24,28 @@ fuzzer_package=${fuzzer_basename}.pkg
...
+    . .venv/bin/activate
+    pip3 install pyyaml
+    export PYTHONPATH="/fuzz-introspector/frontends/python/PyCG"
+
+    ARGS="--fuzzer $fuzzer_path"
+    if [ -n "${PYFUZZPACKAGE-}" ]; then
+      ARGS="$ARGS --package=${PYFUZZPACKAGE}"
+      ARGS="$ARGS --scan"
+    fi
+    python /fuzz-introspector/frontends/python/main.py $ARGS
```
Comment (Navidem) [inline infra/base-images/base-builder/compile_python_fuzzer:44]: Does not go beyond this once `pycg` is install, here is the error:
```
INFO:__main__:Starting analysis
INFO:__main__:Running analysis with arguments: {fuzzer: /src/fuzz_glom.py, package: /src/glom/glom/ }
Traceback (most recent call last):
  File "/fuzz-introspector/frontends/python/main.py", line 354, in <module>
    sys.exit(main())
  File "/fuzz-introspector/frontends/python/main.py", line 342, in main
    exit_code = run_fuzz_pass(
  File "/fuzz-introspector/frontends/python/main.py", line 100, in run_fuzz_pass
    formatter = formats.Fuzz(cg)
AttributeError: module 'pycg.formats' has no attribute 'Fuzz'
```

Comment (Navidem) [reply infra/base-images/base-builder/compile_python_fuzzer:44]: Checking `pycg` [repo](https://github.com/vitsalis/PyCG), there is no such `pycg.formats.Fuzz`.  Maybe it is your own developed module?

**Review by Navidem — CHANGES_REQUESTED** [review]
Now blocked on `base-clang/Dockerfile`.

File: infra/base-images/base-builder/compile_python_fuzzer (hunk 2 above)
Comment (DavidKorczynski) [reply infra/base-images/base-builder/compile_python_fuzzer:44]: Did you rebiuld `base-clang`? This is needed becaues you need the new version of fuzz-introspector that links to the custom PyCG -- see https://github.com/ossf/fuzz-introspector/tree/main/frontends/python

Comment (Navidem) [reply infra/base-images/base-builder/compile_python_fuzzer:44]: Ah! that's the missing step, I just rebuilt `base-builder` and `base-builder-python`. Thanks for prompt response!

**Review by Navidem — CHANGES_REQUESTED** [review]
Once `base-clang/Dockerfile` fixed, it successfully finishes.

File: infra/base-images/base-builder/compile_python_fuzzer [hunk 3]
```diff
@@ -24,6 +24,28 @@ fuzzer_package=${fuzzer_basename}.pkg
...
+if [[ $SANITIZER = *introspector* ]]; then
+    # We must ensure python3.9, this is because we use certain
+    # AST logic from there.
+    # The below should probably be refined
+    apt install -y python3.9
+    apt-get install -y python3-pip
+    python3.9 -m pip install virtualenv
+    python3.9 -m virtualenv .venv
+    . .venv/bin/activate
+    pip3 install pyyaml
```
[the remark this answers is not in the document]
Comment (Navidem) [reply infra/base-images/base-builder/compile_python_fuzzer:36]: nm.

File: infra/base-images/base-clang/Dockerfile [hunk 4]
```diff
@@ -38,7 +38,9 @@ RUN apt-get update && apt-get install -y wget sudo && \
 RUN apt-get update && apt-get install -y git && \
     git clone https://github.com/ossf/fuzz-introspector.git fuzz-introspector && \
     cd fuzz-introspector && \
-    git checkout b0e2a4327197b0064ba28892489423013148bd0d && \
+    git checkout efdcb3c5c49cce2a08fe1856b61f70e3324147e4 && \
+    git submodule --init && \
```
[the remark this answers is not in the document]
Comment (DavidKorczynski) [reply infra/base-images/base-clang/Dockerfile]: Thanks a lot!

File: infra/base-images/base-clang/Dockerfile [hunk 5]
```diff
@@ -38,7 +38,9 @@ RUN apt-get update && apt-get install -y wget sudo && \
...
+    git submodule --update && \
```
[the remark this answers is not in the document]
Comment (DavidKorczynski) [reply infra/base-images/base-clang/Dockerfile]: Thanks a lot!

**Comment by Navidem** [comment]
/gcbrun trial_build.py glom idna pyyaml --sanitizer introspector --fuzzing-engine libfuzzer

File: infra/base-images/base-builder/compile (hunk 1 above)
Comment (DavidKorczynski) [reply infra/base-images/base-builder/compile]: Done!

**Comment by DavidKorczynski** [comment]
Is this ready for merge?

**Comment by Navidem** [comment]
/gcbrun trial_build.py glom idna pyyaml --sanitizer introspector --fuzzing-engine libfuzzer --force

**Comment by Navidem** [comment]
> Is this ready for merge?

Looks good to me, I would wait for trial build to succeed.

**Comment by DavidKorczynski** [comment]
The change bumps fuzz-introspector, and thus does impact base-clang now that fuzz-introspector is in the main clang. Does it make sense to test more than the three Python projects just to be extra cautious?

**Comment by Navidem** [comment]
> The change bumps fuzz-introspector, and thus does impact base-clang now that fuzz-introspector is in the main clang. Does it make sense to test more than the three Python projects just to be extra cautious?

Good point! Let's wait for the python ones to finish then I run another trial.

**Comment by Navidem** [comment]
/gcbrun trial_build.py htslib bzip2 aspell boringssl --sanitizer introspector --fuzzing-engine libfuzzer --force

**Comment by Navidem** [comment]
All trials finished successfully. Good to merge.

**Comment by DavidKorczynski** [comment]
> All trials finished successfully. Good to merge.

Awesome. I can't merge this myself but am happy to get this in.

**Pull request merged**